src/WebJobs.Script/Workers/Http/HttpWorkerFunctionProvider.cs

```diff
@@ -0,0 +1,149 @@
+            ArgumentNullException.ThrowIfNull(httpWorkerOptions?.Value);
+            ArgumentNullException.ThrowIfNull(languageWorkerOptions);
+            ArgumentNullException.ThrowIfNull(hostFunctionMetadataProvider);
+            _httpWorkerOptions = httpWorkerOptions?.Value;
+            _hostFunctionMetadataProvider = hostFunctionMetadataProvider;
+            _languageWorkerOptions = languageWorkerOptions;
+            _logger = logger;
+        }
+
+        public ImmutableDictionary<string, ImmutableArray<string>> FunctionErrors => _errors;
+
+        private static JArray BuildAllHttpMethods()
+        {
+            return new JArray(
+                HttpMethods.Get,
+                HttpMethods.Post,
+                HttpMethods.Put,
+                HttpMethods.Delete,
+                HttpMethods.Patch,
+                HttpMethods.Head,
+                HttpMethods.Options,
+                HttpMethods.Trace,
+                HttpMethods.Connect);
+        }
+
+        public async Task<ImmutableArray<FunctionMetadata>> GetFunctionMetadataAsync()
+        {
+            var routes = _httpWorkerOptions.Http?.Routes;
+
+            if (routes is null || !routes.Any())
+            {
+                return [];
+            }
+
+            if (!_httpWorkerOptions.CustomRoutesEnabled)
+            {
+                throw new InvalidOperationException($"Routes configuration is only allowed for worker runtime: custom");
+            }
+
+            var hostFunctionMetadata = await _hostFunctionMetadataProvider.GetFunctionMetadataAsync(_languageWorkerOptions.CurrentValue.WorkerConfigs, forceRefresh: false);
+
+            // We already know custom handler http routes are configured, if function.json files are also present we cannot proceed.
+            if (hostFunctionMetadata.Any())
+            {
+                throw new InvalidOperationException(
+                    "Detected both function.json files and custom handler HTTP route configuration definition in host.json" +
+                    "Only one configuration source is supported. Remove either the function.json files or the HTTP routes entries in host.json.");
+            }
+
+            return CreateFunctionsFromRoutes(routes);
+        }
+
+        private ImmutableArray<FunctionMetadata> CreateFunctionsFromRoutes(IEnumerable<HttpWorkerRoute> routes)
+        {
+            var metadatabuilder = ImmutableArray.CreateBuilder<FunctionMetadata>();
+            var errorsBuilder = ImmutableDictionary.CreateBuilder<string, ImmutableArray<string>>();
+            int i = 0;
+
+            foreach (var route in routes)
+            {
+                var functionName = $"http-handler{++i}";
+
+                try
+                {
+                    // Template parser does not check for empty route.
+                    if (string.IsNullOrEmpty(route.Route))
+                    {
+                        throw new ArgumentException("Route cannot be null or empty.");
```

Comment:
This is not a valid exception here. The individual route is not an argument of this method.